File: Go/maximum-score-from-removing-substrings.go
func maximumGain(s string, x int, y int) int {
    points, count := 0, 0
    if x >= y {
        s, count = removeAndCount(s, "ab")
        points += x * count
        _, count = removeAndCount(s, "ba")
        points += y * count
    } else {
        s, count = removeAndCount(s, "ba")
        points += y * count
        _, count = removeAndCount(s, "ab")
        points += x * count
    }
    return points
}

func removeAndCount(s string, sub string) (string, int) {
    count := 0
    stack := make([]byte, len(s))
    top := -1
    for _, c := range s {
        if top >= 0 && byte(c) == sub[1] && stack[top] == sub[0] {
            count++
            top--
        } else {
            top++
            stack[top] = byte(c)
        }
    }    
    return string(stack[:top+1]), count
}

File: C++/maximum-score-from-removing-substrings.cpp
class Solution {
public:
    int maximumGain(string s, int x, int y) {
        int points = 0;
        int count = 0;
        if (x >= y) {
            count = removeAndCount(s, "ab");
            points += x * count;
            count = removeAndCount(s, "ba");
            points += y * count;
        } else {
            count = removeAndCount(s, "ba");
            points += y * count;
            count = removeAndCount(s, "ab");
            points += x * count;
        }
        return points;
    }

    int removeAndCount(string& s, string sub) {
        int count = 0;
        vector<char> stack;
        int top = -1;
        for (auto c : s) {
            if (!stack.empty() && c == sub[1] && stack.back() == sub[0]) {
                count++;
                stack.pop_back();
            } else {
                stack.push_back(c);
            }
        }
        string tmp(stack.begin(), stack.end());
        s = tmp;
        return count;
    }
};